include/xparrow/xbuffer.hpp

```diff
@@ -0,0 +1,239 @@
+
+#include <algorithm>
+#include <concepts>
+#include <cstdint>
+
+namespace xp
+{
+    namespace impl
+    {
+        template <class T>
+        struct xbuffer_data
+        {
+            using value_type = T;
+            using pointer = T*;
+            using size_type = std::size_t;
+            
+            bool empty() const noexcept;
+            size_type size() const noexcept;
+
+            template <class U = T>
+            U* data() noexcept;
+
+            template <class U = T>
+            const U* data() const noexcept;
+
+            void swap(xbuffer_data& rhs) noexcept;
+            bool equal(const xbuffer_data& rhs) const;
+
+            pointer p_data = nullptr;
+            size_type m_size = 0;
+        };
+    }
+
+    /**
+     * @class xbuffer
+     * @brief Object that owns a piece of contiguous memory
+     */
+    template <class T>
+    class xbuffer : private impl::xbuffer_data<T>
+    {
+    public:
+
+        using base_type = impl::xbuffer_data<T>;
+        using value_type = typename base_type::value_type;
+        using pointer = typename base_type::pointer;
+        using size_type = typename base_type::size_type;
+
+        xbuffer() = default;
+        explicit xbuffer(size_type size);
+        xbuffer(pointer data, size_type size);
+        
+        ~xbuffer();
+        
+        xbuffer(const xbuffer&);
+        xbuffer& operator=(const xbuffer&);
+
+        xbuffer(xbuffer&&);
+        xbuffer& operator=(xbuffer&&);
+
+        using base_type::empty;
+        using base_type::size;
+        using base_type::data;
+
+        void resize(size_type new_size);
+
+        void swap(xbuffer&) noexcept;
+        bool equal(const xbuffer& rhs) const;
+
+    private:
+
+        pointer allocate(size_type size) const;
+        void deallocate(pointer mem) const;
+    };
+
+    template <class T>
+    bool operator==(const xbuffer<T>& lhs, const xbuffer<T>& rhs);
+
+    /**
+     * @class xbuffer_reference
+     * @brief Object that wraps a piece of contiguous memory
+     * but does not own it.
+     */
+    /*class xbuffer_reference : private xbuffer_data
+    {
+    public:
+
+        using base_type = xbuffer_data;
+
+        xbuffer_reference(uint8_t* data, std::size_t size);
+
+        using base_type::size;
+        using base_type::data;
+        
+        void swap(xbuffer_reference&) noexcept;
+    };*/
+
+    /******************************
+     * xbuffer_data implementation *
+     ******************************/
+
+    namespace impl
+    {
+        template <class T>
+        bool xbuffer_data<T>::empty() const noexcept
+        {
+            return size() == size_type(0);
+        }
+
+        template <class T>
+        auto xbuffer_data<T>::size() const noexcept -> size_type
+        {
+            return m_size;
+        }
+
+        template <class T>
+        template <class U>
+        U* xbuffer_data<T>::data() noexcept
+        {
+            return reinterpret_cast<U*>(p_data);
```

Comment:
shouldnt `static_cast` make more sense here? it's not like it's converting types to/from `void*` so both types needs to have some form of compatibility.

Reply:
We might convert from / to `uint8_t*` if we store buffer of raw memory.

Reply:
I would prefer a specific operator for that then, something like `raw_data()` or `bytes()`. Cause right now it's allowed to convert to anything without any warning.

Reply:
That would handle the case for converting *to* `uint8_t`, but we may also want to convert *from* `uint8_t`, to any unrelated type.

Reply:
I dont think I understand what you mean here?

Reply:
we may use `xbuffer<uint8_t>` as a storage for raw memory. Then, in an array class that depends on it, we know (from somewhere) that's the actual data type is `double`, so we may want to do something like:
```cpp
double* dagta_buffer = buffer.data<double>();
```

The actual type could be any integral type, or floating point, otr whatever else.

Reply:
I am not convinced that this is useful as it hides a potentially dangerous cast in an apparently safe call.
I'm not against going with it and seeing where it leads, but my gut feeling is that it will hide bugs.